src/cargo/core/compiler/fingerprint/mod.rs

```diff
@@ -796,14 +798,36 @@ pub enum StaleItem {
 impl LocalFingerprint {
     /// Read the environment variable of the given env `key`, and creates a new
     /// [`LocalFingerprint::RerunIfEnvChanged`] for it.
-    ///
-    // TODO: This is allowed at this moment. Should figure out if it makes
-    // sense if permitting to read env from the config system.
     #[allow(clippy::disallowed_methods)]
-    fn from_env<K: AsRef<str>>(key: K) -> LocalFingerprint {
+    fn from_env<K: AsRef<str>>(
+        key: K,
+        env_config: &Arc<HashMap<String, OsString>>,
+        env_config_insensitive: &Arc<LazyCell<HashMap<String, OsString>>>,
+    ) -> LocalFingerprint {
         let key = key.as_ref();
         let var = key.to_owned();
-        let val = env::var(key).ok();
+        let val = if let Some(val) = env::var(key).ok() {
+            Some(val)
+        } else {
+            match env_config.get(key) {
+                Some(value) => value.to_str().map(|s| s.to_string()),
+                None => {
+                    if cfg!(windows) {
+                        env_config_insensitive
+                            .borrow_with(|| {
+                                env_config
+                                    .iter()
+                                    .map(|(k, v)| (k.to_uppercase().clone(), v.clone()))
+                                    .collect::<HashMap<String, OsString>>()
+                            })
+                            .get(&key.to_uppercase())
+                            .and_then(|s| s.to_str().map(|s| s.to_string()))
```

Comment:
Please use a more specific conversion than `to_string` as it can mask incidental semantic conversions

Reply:
I guessed you intention is explictly to use `str::to_string` instead ?

Reply:
For myself, I tend to prefer `to_owned` as that signals the intent